<!-- # Child Rmarkdown Document for 10X genomics cellranger-based scRNA sequencing outputs -->
<!-- Parent document must have a variable "in_rna" containing metadata-injected H5 files by well --> 
<a id="rna_seq_well_top"></a>

```{r, include=FALSE} 
orig_workspace <- ls()

scrna_seq_well_module_version <- "1.0.2" # 20211201
stm("Starting scRNA Well Module")
```

The following metrics summarize the sequencing and alignment by 10x well prior to un-hashing and hash-based cell filtering.  

### Contents 

#### [Pool Summary](#pool_seq_summary)
#### [Detailed Well Summary](#well_seq_summary)
#### [Plots of Well-Level Metrics](#seq_metrics_plots)  
#### [Count Stats per Well](#rna_well_counts)  
  - [Read Counts per Well Plot](#rna_read_count)  
  - [UMI Counts per Well Plot](#rna_umi_count)  
  - [Gene Counts per Well Plot](#rna_gene_count)  
  
#### [Mitochondrial UMIs](#mito_umis_rna)
  - [Fraction Mitochondrial UMI](#rna_fct_mito)  
  - [Fraction MT UMI by UMI Counts](#rna_fct_mito_by_umi_counts)  
  - [Fraction MT UMI by Gene Counts](#rna_fct_mito_by_gene_counts)  
    
#### [Data Quality UMAP](#rna_umap)  

<details style="color: lightgray;">  
  <summary>Expand Code</summary> 
  
Check Dependencies  
```{r scrna_seq_dependency_check, include = FALSE}
assertthat::assert_that(exists("in_rna"))  

all_h5 <- list.files(path = in_rna, 
                               pattern = ".h5$", 
                               full.names = TRUE, recursive = TRUE)
assertthat::assert_that(length(all_h5) >0, 
                        msg = sprintf("Did not detect any .h5 files in input RNA directory %s", in_rna))

```   

Reading in well info from h5 files
```{r scrna_seq_input_files}

if(!exists("well_info")){
  # Merge Well Data
  stm("Reading in h5 file well data")
  well_list <- lapply(all_h5, read_h5_well_meta)
  
  # Find common columns
  cols_keep <- Reduce(intersect, lapply(well_list, colnames))
  well_list <- lapply(well_list, function(x){x[,cols_keep]})
  
  well_info <- unique(do.call(rbind, well_list))
  setDT(well_info)
  well_info[, pool_id := gsub("C\\d+W\\d+", "", well_id)]
  
  remove("well_list")
}

``` 

Reading in metadata from h5 files  
```{r rna_metadata}
stm("Reading and merging all rna meta data")
rna_meta_list <- lapply(all_h5, H5weaver::read_h5_cell_meta)

# make sure column names are the same
col_list <- lapply(rna_meta_list, colnames)
all_cols_identical <- length(unique(col_list)) == 1
if(!all_cols_identical){
  all_columns <- unique(unlist(lapply(rna_meta_list, colnames)))
  common_columns <- Reduce(union, lapply(rna_meta_list, colnames))
  if(!all(all_columns %in% common_columns)){
    stm(sprintf("Warning: rna h5 files do not contain the same meta data columns. Keeping only the common columns. Removing columns %s.", 
        paste(setdiff(all_columns, common_columns), sep = ", ")))
  }
  rna_meta_list <- lapply(rna_meta_list, function(x){x[, common_columns]})
}

# merge metadata
rna_meta <- do.call(rbind, rna_meta_list)

# add metadata variables
rna_meta$fct_mito_umi <- rna_meta$n_mito_umi/rna_meta$n_umis
fct_mito_grp_cutoffs <- c(-Inf, 0.05, 0.10, 0.20, 0.30,Inf)
fct_mito_grp_labels <- c("0-0.05","0.05-0.10","0.10-0.20","0.20-0.30",">0.30")
rna_meta$fct_mito_group <- cut(rna_meta$fct_mito_umi, breaks =fct_mito_grp_cutoffs,
                           labels = fct_mito_grp_labels)
```  

Read in Counts from H5 Files  
```{r rna_counts}
stm("Reading and merging all rna count matrices")

rna_count_list <- lapply(all_h5, H5weaver::read_h5_dgCMatrix, target = "matrix", 
                         feature_names = "id")
# make sure all matrices have same number of rows
if(!length(unique(sapply(rna_count_list, nrow)))==1){
  stop("RNA count matrixes have different numbers of rows")
} 

# make sure rows are in same order
row_order <- rownames(rna_count_list[[1]])
rna_count_list <- lapply(rna_count_list, function(x){x[row_order,]})

# make sure columns are in same order as metadata
order_check <- mapply(function(x, y){(all(x$barcodes==colnames(y)))}, rna_meta_list, rna_count_list)
if(!all(order_check)){
  # Reorder matrix columns to be consistent with metadata
  rna_count_list <- mapply( function(x, y){x[,y$barcodes]}, rna_count_list, rna_meta_list)
}

# merge
rna_counts <- do.call(cbind, rna_count_list)

featDF <- read_h5_feature_meta(all_h5[1])
``` 

Sample cells from each well to generate umaps  
```{r rna_sampled_cells}
n_cells_sample <- 2000

stm(sprintf("Sampling %s cells per well (or all cells if fewer)", n_cells_sample))

set.seed(3)
sample_index_list <- lapply(rna_meta_list, function(x){
  sample_size = min(n_cells_sample, nrow(x))
  sort(sample(1:nrow(x), size = sample_size, replace = FALSE))
})

n_files <- length(sample_index_list)
rna_meta_list_sampled <- lapply(1:n_files, function(x){
 rna_meta_list[[x]][sample_index_list[[x]],]
})
rna_count_list_sampled <- lapply(1:n_files, function(x){
 rna_count_list[[x]][,sample_index_list[[x]]]
})

rna_meta_sampled <- do.call(rbind, rna_meta_list_sampled)
rna_meta_sampled$fct_mito_umi <- rna_meta_sampled$n_mito_umi/rna_meta_sampled$n_umis
rna_meta_sampled$fct_mito_group <- cut(rna_meta_sampled$fct_mito_umi, breaks =fct_mito_grp_cutoffs,
                           labels = fct_mito_grp_labels)
rna_counts_sampled <- do.call(cbind, rna_count_list_sampled)

rm(list=c("rna_count_list", "rna_count_list_sampled","rna_meta_list"))

```  

```{r scrna_definitions} 
vnames_rna <- c("estimated_number_of_cells", "fraction_reads_in_cells", 
            "mean_reads_per_cell", "median_genes_per_cell", "median_umi_counts_per_cell", 
            "number_of_reads", "q30_bases_in_barcode", "q30_bases_in_rna_read", 
            "q30_bases_in_sample_index", "q30_bases_in_umi", "reads_mapped_antisense_to_gene", 
            "reads_mapped_confidently_to_exonic_regions", "reads_mapped_confidently_to_genome", 
            "reads_mapped_confidently_to_intergenic_regions","reads_mapped_confidently_to_intronic_regions", 
            "reads_mapped_confidently_to_transcriptome", "reads_mapped_to_genome", 
            "sequencing_saturation", "total_genes_detected", "valid_barcodes")

vnames_arc <- c("estimated_number_of_cells",
  "gex_fraction_of_transcriptomic_reads_in_cells","gex_mean_raw_reads_per_cell","gex_median_genes_per_cell",
  "gex_median_umi_counts_per_cell","gex_percent_duplicates","gex_q30_bases_in_barcode",
  "gex_q30_bases_in_read_2","gex_q30_bases_in_sample_index_i1","gex_q30_bases_in_sample_index_i2",
  "gex_q30_bases_in_umi","gex_reads_mapped_antisense_to_gene","gex_reads_mapped_confidently_to_exonic_regions",
  "gex_reads_mapped_confidently_to_genome","gex_reads_mapped_confidently_to_intergenic_regions", "gex_reads_mapped_confidently_to_intronic_regions",
  "gex_reads_mapped_confidently_to_transcriptome","gex_reads_mapped_to_genome","gex_reads_with_tso",
  "gex_sequenced_read_pairs","gex_total_genes_detected","gex_valid_barcodes", "gex_valid_umis")  
  
  
vnames_tcr<-c('cells','confidently mapped antisense','confidently mapped antisense 1',
              'confidently mapped to exonic regions','confidently mapped to exonic regions 1',
              'confidently mapped to genome','confidently mapped to genome 1',
              'confidently mapped to intergenic regions','confidently mapped to intergenic regions 1',
              'confidently mapped to intronic regions','confidently mapped to intronic regions 1',
              'confidently mapped to transcriptome','confidently mapped to transcriptome 1',
              'estimated number of cells','fraction reads in cells','mapped to genome',
              'mapped to genome 1','mean reads per cell','median genes per cell',
              'median reads per cell','median umi counts per cell',
              'number of reads','number of reads 1',
              'number of reads assigned to the sample','number of reads in the library',
              'number of short reads skipped','q30 barcodes','q30 rna read','q30 umi',
              'sequencing saturation','total genes detected','valid barcodes','valid umis')

if(all(vnames_rna %in% colnames(well_info))){
  vnames <- vnames_rna
  vlabels <- c("Estimated Number of Cells", "Fraction Reads in Cells", 
            "Mean Reads per Cell", "Median Genes per Cell", "Median UMI per Cell", 
            "Number of Reads", "Q30 Bases in Barcode (%)", "Q30 Bases in RNA Read (%)", 
            "Q30 Bases in Sample Index (%)", "Q30 Bases in UMI (%)", "Reads Mapped Antisense to Gene (%)", 
            "Reads Mapped Confidently to Exonic Regions (%)", "Reads Mapped Confidently to Genome (%)", 
            "Reads Mapped Confidently to Intergenic Regions (%)",
            "Reads Mapped Confidently to Intronic Regions (%)", 
            "Reads Mapped Confidently to Transcriptome (%)", "Reads Mapped to Genome (%)", 
            "Sequencing Saturation (%)", "Total Genes Detected", "Valid Barcodes (%)")
  n_vars <- length(vnames)
  vartypes <- c(rep("Cells", 5), rep("Sequencing", 5), rep("Mapping", 7),"Sequencing","Cells","Sequencing")
  vartypes <- factor(vartypes,  levels = c("Cells", "Sequencing", "Mapping"))

  digitsRound <- c(0, 1, rep(0, 4), rep(1, 12), 0, 1)
  rna_data_type <- "rna"
} else if (all(vnames_arc %in% colnames(well_info))){
  vnames <- vnames_arc
  vlabels <- c("Estimated Number of Cells","Fraction Transcriptomic Reads in Cells",
            "Mean Raw Reads per Cell", "Median Genes per Cell", "Median UMI per Cell", "Percent Duplicates",
            "Q30 Bases in Barcode", "Q30 Bases in RNA Read 2", "Q30 Bases in Sample Index i1",
            "Q30 Bases in Sample Index i2", "Q30 Bases in UMI", "Reads Mapped Antisense to Gene", 
            "Reads Mapped Confidently to Exonic Regions", "Reads Mapped Confidently to Genome", 
            "Reads Mapped Confidently to Intergenic Regions", 
            "Reads Mapped Confidently to Intronic Regions", 
            "Reads Mapped Confidently to Transcriptome", "Reads Mapped to Genome", "Reads with TSO",
            "Sequenced Read Pairs", "Total Genes Detected", "Valid Barcodes", "Valid UMIs") 
  n_vars <- length(vnames)
  vartypes <- c(rep("Cells", 5), rep("Sequencing", 6), rep("Mapping", 7),"Sequencing","Sequencing","Cells","Sequencing","Sequencing")
  digitsRound <- c(0, 3, 0, 0, 0, rep(3, 14), 0, 0, 3, 3)
  

    rna_data_type <- "arc"
} else if (all(vnames_tcr %in%  gsub("_", " ", colnames(well_info)))){
  vnames <- gsub(" ", "_", vnames_tcr)
  vlabels <- c('Cells','Confidently Mapped Antisense in Sample (%)','Confidently Mapped Antisense in Library (%)',
          'Confidently Mapped to Exonic Regions in Sample (%)','Confidently Mapped to Exonic Regions in Library (%)',
          'Confidently Mapped to Genome (%) in Sample','Confidently Mapped to Genome (%)',
          'Confidently Mapped to Intergenic regions (%) in Sample','Confidently Mapped to Intergenic regions in Library  (%)',
          'Confidently Mapped to Intronic regions (%) in Sample','Confidently Mapped to Intronic regions in Library (%)',
          'Confidently Mapped to Transcriptome (%) in Sample','Confidently Mapped to Transcriptome in Library (%)',
          'Estimated Number of Cells',
          'Fraction Reads in Cells (%)',
          'Mapped to Genome in Sample','Mapped to Genome in Library',
          'Mean Reads per Cell','Median Genes per Cell',
          'Median Reads per Cell','Median UMI Counts per Cell','Number of Reads in Sample','Number of Reads in Library',
          'Number of Reads Assigned to the Sample','Number of Reads in the Library','Number of Short Reads Skipped',
          'Q30 Barcodes (%)','Q30 RNA Read (%)','Q30 UMI (%)','Sequencing Saturation (%)',
          'Total Genes Detected','Valid Barcodes (%)','Valid UMIs')
          
  n_vars <- length(vnames)
  vartypes <- c(rep("Cells", 1), rep("Mapping", 12), "Cells","Cells",rep("Sequencing", 2),rep("Cells", 4),rep("Sequencing", 12))
  digitsRound <- c(0,  rep(3, 12), 0, 0, 3, 3,rep(0, 4),rep(3, 12))
  

    rna_data_type <- "tcr"
}else {
    vnames <- setdiff( colnames(well_info), c("well_id","pool_id"))
    n_vars <- length(vnames)
    vlabels <- gsub("_"," ", vnames)
    vartypes <- rep(NA, n_vars)
    
    getDigits <- function(vNumeric){
      vChar <- as.character(vNumeric)
      is_decimal <- any(grepl("[.]", vChar))
      if(is_decimal){
        decimal_digits <- nchar(vChar) - (nchar(gsub("[.].*","", vChar))+1)
        n_digits <- max(decimal_digits)
      } else {
        n_digits <- 0
      }
      n_digits
    }
    digitsRound <- sapply(subset(well_info, select = vnames), getDigits)
    
    rna_data_type <- "unknown"
}

  df_vars <- data.frame(Category = vartypes,
                        Variable_name = vlabels,
                        Variable = vnames,
                        Round = digitsRound)

```  
  
</details>  

<a  id="pool_seq_summary"></a> 


### Pool Summary    
Summary by pool or by entire non-hashed batch   

```{r well_summary_table1, results = "asis"}
stm("Creating scrna well cellranger summary table")

if(rna_data_type == "rna"){
  pool_info <- well_info %>%
  dplyr::group_by(pool_id) %>%
  dplyr::summarize(total_cells = formatC(sum(estimated_number_of_cells), big.mark = ",", digits = 0, format = "f"),
            total_reads = formatC(sum(number_of_reads), big.mark = ",", digits = 0, format = "f"), .groups = "drop")
} else if (rna_data_type == "arc"){
  pool_info <- well_info %>%
  dplyr::group_by(pool_id) %>%
  dplyr::summarize(total_cells = formatC(sum(estimated_number_of_cells), big.mark = ",", digits = 0, format = "f"),
            total_sequenced_read_pairs = formatC(sum(gex_sequenced_read_pairs), big.mark = ",", digits = 0, format = "f"), .groups = "drop")
} else if (rna_data_type == "tcr"){
  pool_info <- well_info %>%
  dplyr::group_by(pool_id) %>%
  dplyr::summarize(total_cells = formatC(sum(estimated_number_of_cells), big.mark = ",", digits = 0, format = "f"),
            total_reads = formatC(sum(number_of_reads), big.mark = ",", digits = 0, format = "f"), .groups = "drop")
}else {
  print("Warning: No total reads column detected in well metadata")
  pool_info <- well_info %>%  
  dplyr::group_by(pool_id) %>%  
  dplyr::summarize(total_cells = formatC(sum(estimated_number_of_cells), big.mark = ",", digits = 0, format = "f"), 
                   total_reads = NA, .groups = "drop")
}

names(pool_info) <- stringr::str_to_title(gsub("_", " ", names(pool_info)))
pool_info %>%
  gt::gt() %>%
  gt::cols_align(align = "right", columns = 2:3)
  
rm(pool_info)
```  
[Return to Contents](#rna_seq_well_top) 

<a  id="well_seq_summary"></a>  
  
### Detailed Well Summary  
```{r well_summary_table, results = "asis"}  
stm("Creating detailied scrna well cellranger table")

unique_pools <- sort(unique(well_info$pool_id))  

well_summary_table <- well_info %>% 
  gather(key = Variable, value = Value, all_of(vnames)) %>%   # all variables long
  full_join(df_vars, by = "Variable") %>%  
  group_by(pool_id, Category, Variable, Variable_name) %>% 
  summarize(Median = formatC(median(Value, na.rm=T), big.mark = ",", digits = unique(Round), format = "f"),
            Range = get_range(Value, digits = unique(Round), verbose = F), 
            `CV%` = round(sd(Value)/mean(Value)*100, 1),
            .groups = "drop") %>% 
  arrange(Category, Variable_name) %>% 
  tidyr::pivot_wider(id_cols = c("Category", "Variable", "Variable_name"), 
                     names_from = pool_id, 
                     values_from = c("Median", "Range", "CV%"),
                     names_glue = "{pool_id}__{.value}",
                     names_sort = TRUE) %>%
  mutate(Plot = sprintf("[Plot](#%s)", Variable)) %>% 
  select(Category, Variable_name, contains(unique_pools), Plot) # reorder cols by pools first then stats, keep only the clean var name

gt_table <- well_summary_table %>% 
  gt::gt() %>% 
  gt::fmt_markdown(columns = "Plot") %>% # convert the plot column text to markdown to activate links. these links generated with plots below.
  gt::cols_width(vars(Category) ~ px(100),
                 vars(Variable_name) ~ px(150),
                 ends_with("Median") ~ px(100),
                 ends_with("Range") ~ px(100),
                 ends_with("CV%") ~ px(50),
                 vars(Plot) ~ px(60)) %>% 
  gt::tab_options(table.font.size = 11, column_labels.font.size = 12) %>%
  gt::tab_spanner_delim(delim = "__") %>% 
  gt::cols_align(align = "right",
                 columns = c(ends_with("Median"), ends_with("Range"))) %>% 
  gt::cols_align(align = "center",
                 columns = vars(Plot)) %>% 
  gt::cols_label(Variable_name = "Variable")

gt_table  

rm(well_summary_table)
rm(gt_table)
  
```  

<details>
  <summary>Expand table of statistics per well</summary> 
```{r well_info_table}
qc_table(well_info)

```  

</details>   

[Return to Contents](#rna_seq_well_top) 

<a  id="seq_metrics_plots"></a> 
  
### Plots of Well-Level Metrics  
  
```{r well_plots, results = "asis", fig.height = 3.5}
stm("Generating sequencing and alignment QC plots")

verpal <- hcl.colors(n = n_vars, palette = "viridis")

# Plots
for (i in seq_along(vnames)){
  df <- data.table::copy(well_info)
  spec <- vnames[i]
  slabel <- vlabels[i]
  df <- as.data.frame(df)
  df$spec_col <- df[,spec]
  med_val <- median(df$spec_col)
  cv <- round(sd(df$spec_col)/mean(df$spec_col)*100, 2)
  n <- sum(!is.na(df$spec_col))
  
  g <- ggplot(df, aes(well_id, spec_col)) +
    geom_bar(stat = "identity", fill = verpal[i]) + 
    geom_hline(yintercept = med_val, linetype = "dashed", color = "red")+
    scale_y_continuous(sec.axis = dup_axis(breaks = med_val, labels = med_val, name = ""))+
    xlab("Well") +
    ylab(slabel) +
    facet_wrap(~pool_id, ncol = n_pools, scales = "free_x", drop = TRUE) +
    ggtitle(slabel, 
            subtitle = sprintf("Median=%s    CV=%.1f%%    N=%s", med_val, cv, n)) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
  
  # Plot-specific hyperlink definition
  cat(sprintf('\n<a id="%s"></a>', spec), labels = "", sep = "\n")
  
  # Output plot
  suppressWarnings(print(g))
  
  # Link back to top of section
  cat("  \n[Return to Contents](#rna_seq_well_top)", labels = "", sep = "\n")
  
  rm(df, spec, slabel, med_val, cv, n, g)
  
}

```  

<a id="rna_well_counts"></a> 

### Count Stats per Well

<a id= "rna_read_count"></a>  

#### Read Counts by Well
```{r read_counts_well, results="asis"}
# Reads per hto cat
stm("Generating read count violin plots")

# Reads per hto plot
g_read <- qc_violin_plot(rna_meta,
                        category_x = "well_id",
                        name_x = "Well",
                        column_y = "n_reads",
                        name_y = "N Reads per Cell",
                        log_y = TRUE,
                        fill = "dodgerblue") +
  ggtitle("Reads per Well")

temp_figwidth = max(5, 0.5 + n_wells*0.4)
temp_figheight = 4

batchreporter::make_subchunk(g_read, subchunk_name = "well_read_violin_subchunk", 
              chunk_opt_list = list(fig.height = temp_figheight, fig.width = temp_figwidth, 
                                    warning = FALSE), 
              quiet_knit = TRUE)
              
rm(g_read)

```  

[Return to Contents](#rna_seq_well_top) 

<a id= "rna_umi_count"></a>  

#### UMI Counts by Well
```{r umi_counts_well, results="asis"}
# Reads per hto cat
stm("Generating umi count violin plots")

# UMI per hto plot
g_umi <- qc_violin_plot(rna_meta,
                        category_x = "well_id",
                        name_x = "Well",
                        column_y = "n_umis",
                        name_y = "N UMIs per Cell",
                        log_y = TRUE,
                        fill = "purple") +
  ggtitle("UMIs per Well")

temp_figwidth = max(5, 0.5 + n_wells*0.4)
temp_figheight = 4

batchreporter::make_subchunk(g_umi, subchunk_name = "well_umi_violin_subchunk", 
              chunk_opt_list = list(fig.height = temp_figheight, fig.width = temp_figwidth, 
                                    warning = FALSE), 
              quiet_knit = TRUE)
rm(g_umi)

```  

[Return to Contents](#rna_seq_well_top) 

<a id= "rna_gene_count"></a>  

#### Gene Counts by Well
```{r gene_counts_well, results="asis"}
# Reads per hto cat
stm("Generating gene count violin plots")

# Reads per hto plot
g_genes <- qc_violin_plot(rna_meta,
                        category_x = "well_id",
                        name_x = "Well",
                        column_y = "n_genes",
                        name_y = "N Genes per Cell",
                        log_y = TRUE,
                        fill = "orangered") +
  ggtitle("Genes per Well")

temp_figwidth = max(5, 0.5 + n_wells*0.4)
temp_figheight = 4

batchreporter::make_subchunk(g_genes, subchunk_name = "well_gene_violin_subchunk", 
              chunk_opt_list = list(fig.height = temp_figheight, fig.width = temp_figwidth, 
                                    warning = FALSE), 
              quiet_knit = TRUE)

rm(g_genes)  

``` 

[Return to Contents](#rna_seq_well_top) 

<a id= "mito_umis_rna"></a>  

### Mitochondrial UMIs

<a id= "rna_fct_mito"></a>  

#### Fraction Mitochondrial UMI  

```{r fct_mito_plot_nohash, results = "asis"}
g_bar <- ggplot(rna_meta, aes(well_id,  fill = factor(fct_mito_group, 
                                                      levels = rev(fct_mito_grp_labels)))) +
  geom_bar() +
  labs(fill = "Fraction Mitochondrial UMIs") +
  ylab("N Cells") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5))

g_pl <- plotly::ggplotly(g_bar) %>%
  layout(hovermode = 'compare')

tempwidth <- n_samples*0.4 + 2

make_subchunk(g_pl, subchunk_name = "fraction_mito_subchunk", quiet_knit = T,
              chunk_opt_list = list(fig.width = tempwidth, fig.height = 6, echo = FALSE))

rm(g_bar, g_pl)  

```  

[Return to Contents](#rna_seq_well_top) 

<a id= "rna_fct_mito_by_umi_counts"></a>  

#### Fraction Mito UMIs by UMI Counts  
```{r rna_mito_umis_vs_total, width = 8, height = 6}
stm("Generating mt umi vs umi count scatter plots")

# Reads per hto plot
g_mito_umi <- ggplot(rna_meta, aes(n_umis, fct_mito_umi)) +
    geom_point(color = "purple", alpha = 0.2) +
    xlab("UMI per Cell (log10 scale)") +
    ylab("Fraction Mitochondrial UMI") +
    scale_x_log10() +
    ggtitle("Fraction Mitochondrial UMI vs UMI per Cell") 
g_mito_umi



```  

<details>  
  <summary> Expand Per-Well Plot </summary>  
```{r rna_mito_umis_vs_umis_well, results="asis"}
stm("Generating mt umi vs umi count scatter plots by well")

n_wells_plot <- length(unique(rna_meta$well_id))
ncols_plot <- 6
nrows_plot <- ceiling(n_wells_plot/ncols_plot)

# Reads per hto plot
g_mito_umi_well <- g_mito_umi +
    facet_wrap(~well_id, ncol = ncols_plot)

temp_figwidth <- 2*min(n_wells_plot, ncols_plot)+0.3
temp_figheight <- 2*nrows_plot + 0.3

make_subchunk(g_mito_umi_well, subchunk_name = "fraction_mito_umis_well_subchunk", quiet_knit = T,
              chunk_opt_list = list(fig.width = temp_figwidth, fig.height = temp_figheight, echo = FALSE))

rm(g_mito_umi, g_mito_umi_well)

```  
</details>   

[Return to Contents](#rna_seq_well_top) 

<a id= "rna_fct_mito_by_gene_counts"></a>  

#### Fraction Mito UMIs by Gene Counts  
```{r rna_mito_umis_vs_genes, height = 6, width = 8}
stm("Generating mt umi vs gene count scatter plots")

# Reads per hto plot
g_mito_gene <- ggplot(rna_meta, aes(n_genes, fct_mito_umi)) +
    geom_point(color = "orangered", alpha = 0.2) +
    xlab("Genes per Cell (log10 scale)") +
    ylab("Fraction Mitochondrial UMI") +
    scale_x_log10() +
    ggtitle("Fraction Mitochondrial UMI vs UMI per Cell") 
g_mito_gene

```  

<details>  
  <summary> Expand Per-Well Plot </summary>  
  
```{r rna_mito_umis_vs_genes_well, results="asis"}
stm("Generating mt umi vs gene count scatter plots by well")

n_wells_plot <- length(unique(rna_meta$well_id))
ncols_plot <- 6
nrows_plot <- ceiling(n_wells_plot/ncols_plot)

# Reads per hto plot
g_mito_gene_well <- g_mito_gene +
    facet_wrap(~well_id, ncol = ncols_plot)

temp_figwidth <- 2*min(n_wells_plot, ncols_plot)+0.3
temp_figheight <- 2*nrows_plot + 0.3

make_subchunk(g_mito_gene_well, subchunk_name = "fraction_mito_genes_well_subchunk", quiet_knit = T,
              chunk_opt_list = list(fig.width = temp_figwidth, fig.height = temp_figheight, echo = FALSE))

rm(g_mito_gene, g_mito_gene_well)

```  
</details>  

[Return to Contents](#rna_seq_well_top) 

<a id="rna_umap"></a>  

### Data Quality UMAP  

<details style="color: lightgray;"> 
  <summary>Expand code</summary>  
  
```{r rna_seurat_normalize}
# Create Seurat object
stm("Creating Seurat object from merged data for scRNA UMAP")
merged_so <- Seurat::CreateSeuratObject(counts = rna_counts_sampled)

# Normalize data
stm("Normalizing data for scRNA UMAP")
merged_so <- Seurat::NormalizeData(object = merged_so,
                                   normalization.method = "LogNormalize",
                                   scale.factor = 10000,
                                   margin = 1)

```  

```{r rna_PCA}
pc_dims <- min(ncol(merged_so) - 1, 50)
# suppressWarnings(future::plan("multiprocess", workers = avail_workers))

stm("Finding Variable Features for scRNA UMAP")
merged_so <- FindVariableFeatures(object = merged_so)

stm("Scaling Data for scRNA UMAP")
merged_so <- ScaleData(object = merged_so, verbose = FALSE)

stm("Running PCA for scRNA UMAP")
merged_so <- RunPCA(object = merged_so, npcs = pc_dims, verbose = FALSE)

```

```{r rna_jackstraw_pc_selection, message = FALSE}
# suppressWarnings(future::plan("multiprocess", workers = avail_workers))

stm("Determining dimensionality via jackstraw method for scRNA UMAP")

labels_order <- rna_meta_sampled$well_id[match(colnames(merged_so),rna_meta_sampled$barcodes)]
names(labels_order) <- colnames(merged_so)

jackstraw_cells <- sample_cells(labels_order, 500, seed = 3030)

jackstraw_so <- merged_so[, jackstraw_cells]

jackstraw_so <- JackStraw(object = jackstraw_so,
                         dims = pc_dims,
                         num.replicate = 50,  
                         verbose = FALSE)
jackstraw_so <- ScoreJackStraw(object = jackstraw_so,
                              dims = 1:pc_dims)

pc_pvals <- jackstraw_so@reductions$pca@jackstraw$overall.p.values[,2]
good_pcs <- pc_pvals < 0.05

nPC <- sum(good_pcs)

pc_var <- Stdev(merged_so, reduction = "pca")^2
total_var <- merged_so@reductions$pca@misc$total.variance
var_selected_pc <- sum(pc_var[good_pcs])/total_var
cumvar_string <- sprintf(fmt = "%.1f", var_selected_pc*100)

stm(sprintf("Selected %s significant pcs via JackStraw, %s%% explained variation for scRNA UMAP", nPC, cumvar_string))

rm(jackstraw_cells, jackstraw_so)

```

```{r rna_merged_umap}
# Run UMAP
stm(sprintf("Running UMAP on selected coordinates for scRNA UMAP"))

merged_so <- Seurat::RunUMAP(merged_so,
                                   dims = c(1:50)[good_pcs],
                                   umap.method = "uwot",
                                   seed.use = 3,
                                   verbose = FALSE)

```


```{r rna_umap_cleanup}
# Get UMAP coordinates
umapDF <- merged_so[["umap"]]@cell.embeddings %>%
          as.data.frame() %>%
          dplyr::rename(UMAP_1_merged = UMAP_1, UMAP_2_merged = UMAP_2) %>%
          tibble::rownames_to_column(var = "barcodes")
umapDF <- merge(umapDF, rna_meta_sampled, by = "barcodes")
rownames(umapDF) <- umapDF$barcodes

umapDF <- umapDF[rownames(merged_so@"meta.data"), , drop = F]

```  
</details>  
<!-- end of details section collapsing multiple no-output code blocks to save space -->  
  
Batch-level UMAP of `r {n_cells_sample}` randomly sampled cells per well using `r {nPC}` principal components (`r {cumvar_string}`% explained variance) selected by jackstraw.  

```{r rna_umap_quality,  results = "asis", fig.height = 20, fig.width = 20, class.output = "superbigimage"}
stm("Plotting Batch scRNA UMAP by Data Quality Metrics")

# Gene scaling
max_genes <- max(8000, rna_meta$n_genes)
max_umi <- max(60000, rna_meta$n_umis)
point_size <- 0.2


# Cell Types
g_base <- ggplot(umapDF, aes(UMAP_1_merged, UMAP_2_merged))

# fraction mitochondrial umi
stm("Plotting Fraction Mito UMAP")
g1 <- plot_umap_report(df = umapDF,
                 x_col="UMAP_1_merged",
                 x_lab = "UMAP 1",
                 y_col = "UMAP_2_merged",
                 y_lab = "UMAP 2",
                 title = "Fraction Mitochondrial UMIs",
                 point_size = point_size,
                 color_col = "fct_mito_umi",
                 scale_color_fun = scale_color_fct_mito)

# N Genes
stm("Plotting N Genes UMAP")
g2 <- plot_umap_report(df = umapDF,
                 x_col="UMAP_1_merged",
                 x_lab = "UMAP 1",
                 y_col = "UMAP_2_merged",
                 y_lab = "UMAP 2",
                 title = "N Genes",
                 point_size = point_size,
                 color_col = "n_genes",
                 scale_color_fun = scale_color_genes(max_genes)
)

# N UMIs
stm("Plotting N UMIs UMAP")
g3 <- plot_umap_report(df = umapDF,
                 x_col="UMAP_1_merged",
                 x_lab = "UMAP 1",
                 y_col = "UMAP_2_merged",
                 y_lab = "UMAP 2",
                 title = "N UMIs",
                 point_size = point_size,
                 color_col = "n_umis",
                 scale_color_fun = scale_color_umis(max_umi)
)


# Wells
stm("Plotting Well UMAP")
cols_well <- H5weaver::varibow(n_colors = length(unique(umapDF$well_id)))

scale_color_well <- function(...){
    scale_color_manual(values = cols_well, ...)
}
g4 <- plot_umap_report(df = umapDF,
                 x_col="UMAP_1_merged",
                 x_lab = "UMAP 1",
                 y_col = "UMAP_2_merged",
                 y_lab = "UMAP 2",
                 title = "Well ID",
                 point_size = point_size,
                 color_col = "well_id",
                 scale_color_fun = scale_color_well
) +
  guides(colour = guide_legend(override.aes = list(size = 3)))


aligned_plots <- cowplot::align_plots(g1, g2, g3, g4, align = "hv", axis = "tblr")  # uniform plot sizing

cowplot::plot_grid(aligned_plots[[1]],
                   aligned_plots[[2]],
                   aligned_plots[[3]],
                   aligned_plots[[4]],
                   ncol = 2)
                   
rm(g1, g2, g3, g4, aligned_plots)

```  

[Return to Contents](#rna_seq_well_top) 
 
---  

scRNA seq report well module v.`r {scrna_seq_well_module_version}`, Lauren Okada  

```{r scrna_seq_well_cleanup, include = FALSE}
module_vars <- setdiff(ls(), orig_workspace)
rm(list=module_vars)

gc()

```