<a id="tcr_hashing_top"></a>   

```{r, include=FALSE}  
orig_workspace <- ls() 

tcr_sample_module_version <- "1.0.0" #202201024
options(warn = -1)

stm("Starting BCR Sample Module")
```

The following metrics summarize the sequencing and alignment by Sample ID after hto-parsing and filtering for singlet cells.  


### Contents 

#### [scTCR Hash-Based Summary](#hto_scTCR_summary)  
   - [scTCR Well Counts per HTO Barcode Plot](#tcr_well_hto_barcode)   
  - [scTCR Well Fraction per HTO Barcode Plot](#tcr_well_fraction_hto_barcode)  
  
  
#### [scTCR Sample Clonotype Summary](#sample_scTCR_summary)  
   - [scTCR Clonotype Size](#tcr_sample_clone_size)
  - [scTCR Sample Unique Clonotype](#unique_clone_type) 
  - [scTCR Sample CDR3 Nucleotide Distribution Summary](#sample_cdr3_nt_distribution)
  - [scTCR Sample Clonotyoe Abundance](#tcr_sample_clonotype_abundance)
  
  
```{r include = FALSE}
stm("Starting scTCR Hashing Module (sample-level)")
```  
  
Check analysis dependencies  
```{r hto_tcr_dependency_check, include = FALSE}  
  
contig_files <- list.files(path = in_tcr, 
                               pattern = "Reformated.csv", 
                               full.names = TRUE, recursive = TRUE)
assertthat::assert_that(length(contig_files) >0, 
                        msg = sprintf("Did not detect any TCR contig files in input tcr directory %s", in_tcr)) 
                        
                      
assertthat::assert_that(exists("in_config"),
                         msg = "Attempting to TCR quality with hto module. Missing `in_config` variable definition")
assertthat::assert_that(file.exists(in_config),
                         msg = sprintf("Attempting to run TCR quality with hto module. Input file `in_config` [%s] does not exist", in_config))                        
                        
# check hash folder exists
assertthat::assert_that(dir.exists(in_hto), 
                        msg = "Attempting to run TCR quality with hto module. Input `in_dir` does not contain 'hto' folder")

assertthat::assert_that(exists("in_key"),
                         msg = "Attempting to run quality with hto module. Missing `in_key` variable definition")
assertthat::assert_that(file.exists(in_key),
                         msg = sprintf("Attempting to TCR quality with hto module. Input file `in_key` [%s] does not exist", in_key))                        
                        
                        
```   




```{r hto_tcr_read_config}
# Config
config_list<- batchreporter::load_config(in_config)

hash_key <- config_list$hash_key
sample_column_name <- config_list$sample_column_name

# Sample Key
df_key <- data.table::fread(in_key)

pools <- unique(df_key$PoolID)
n_pools <- length(pools)

wells <- unlist(strsplit(unique(df_key$WellID), split = ";"))
n_wells <- length(wells)

samples <- unique(df_key$SampleID)
n_samples <- length(samples)

```

```{r tcr_contig_reading}

dt<-do.call(rbind,lapply(contig_files,read.csv))

```



```{r hto_tcr_hash_files}
stm("Reading in hto key for hashed TCR analysis")
# Read in all hto key to match hto names to barcodes
hto_key <- system.file(file.path("reference",hash_key), package = "HTOparser")  # Parameterize this value
in_hto_key <- fread(hto_key, header = FALSE, col.names = c("hto_barcode","hto_name")) %>% 
  mutate(hto_order = as.numeric(gsub("HT","", hto_name))) %>% 
  mutate(hto_name = factor(hto_name, levels = hto_name[order(hto_order)])) %>% # use HT number value to reorder the HT levels
  select(-hto_order)

# Read in all hto metadata files, check expected numbers vs input well info
all_hto_meta <- list.files(path = in_hto, 
                           pattern = "hto_category_table.csv.gz", 
                           full.names = TRUE, recursive = TRUE)
n_hto_meta <- length(all_hto_meta)
if(n_hto_meta == 0){
  stop(sprintf("No 'hto_category_table.csv.gz' files found in %s", in_hto))
} else if (n_hto_meta < n_wells){
  hto_meta_warn <- sprintf("Input number of 'hto_category_table.csv.gz' files (%s) is fewer than number of wells (%s) in sample key",
                           n_json, n_wells)
  warning(hto_meta_warn)
  hto_warning_list <- c(hto_warning_list, hto_meta_warn)
}
stm(paste0("IN HTO Metadata Files        :\n\t", paste(all_hto_meta, collapse = "\n\t")))
cat(paste0("IN HTO Metadata Files        :\n\t", paste(all_hto_meta, collapse = "\n\t"))) 

stm("Reading in hto category metadata for hashed TCR analysis")
hto_meta_list <- lapply(all_hto_meta, fread)
hto_meta_wells <- gsub("_.*","",basename(all_hto_meta))
hto_meta_list <- mapply(function(x,y){x$well_id <- y; x}, hto_meta_list, hto_meta_wells, SIMPLIFY = F)
hto_meta <- do.call(rbind, hto_meta_list)
rm("hto_meta_list")

hto_meta[in_hto_key, on = 'hto_barcode', hto_name := i.hto_name]  # merge in the hto names
hto_meta[ , pool_id := gsub("C\\dW\\d","", well_id)]
hto_meta[ , sample_hto:= sprintf("%s\n%s", hto_name, get(sample_column_name))]
hto_meta[ , sample_hto_pool:= sprintf("%s\n%s%s", hto_name, get(sample_column_name), pool_id)]
hto_meta[ , hto_order:=  as.numeric(gsub("HT","", hto_name))]
hto_meta[ , sample_hto_pool:=  factor(sample_hto_pool, levels = unique(sample_hto_pool[order(pool_id, hto_order)]))]
hto_meta[ , hto_order:= NULL]
hto_meta[ , hto_category:= factor(hto_category, levels = c("no_hash", "singlet", "doublet", "multiplet"))]

```   


```{r hto_tcr_merge}
stm("Merging hto category metadata with tcr contig data")
dt$well_id<-paste0(dt$Batch_ID,"-",dt$Well_ID)
dt$cell_barcode<-substr(dt$barcode, 0,16)
dt<-left_join(dt,hto_meta,by=c("cell_barcode","well_id"))
dt<-dt %>% filter (hto_category=="singlet")
unqiue_dt <- unique(dt[ , c("cell_barcode", "well_id")]) 
unqiue_dt<-left_join(unqiue_dt,hto_meta,by=c("cell_barcode","well_id"))
``` 



<a id= "hto_scTCR_summary"></a>  


### scTCR Hash-Based Summary

```{r hto_tcr_hash_singlet_summary}
stm("Output pool based summary table")

Sample_HTO_Summary<-dt %>%    group_by ( pool_id,hto_name,pbmc_sample_id, hto_barcode,barcode,well_id) %>%   
        tally()  %>%group_by ( pool_id,pbmc_sample_id,hto_name, hto_barcode)%>%   
        tally()
setorder(Sample_HTO_Summary, pool_id, hto_name)
chain<-dt %>%    group_by ( pool_id,hto_name, hto_barcode,barcode,well_id,chain) %>%   
        tally()  %>%group_by ( pool_id,hto_name, hto_barcode,chain)%>%   
        tally()
TRA<-as.data.frame(chain) %>% filter (chain=="TRA")  %>% select (hto_name,n)
TRB<-as.data.frame(chain) %>% filter (chain=="TRB")%>% select (hto_name,n)
colnames(TRA)[2]<-"TRA_counts"
colnames(TRB)[2]<-"TRB_counts"
Sample_HTO_Summary<-left_join(Sample_HTO_Summary,TRA,by=c("hto_name"))
Sample_HTO_Summary<-left_join(Sample_HTO_Summary,TRB,by=c("hto_name"))
colnames(Sample_HTO_Summary)[5]<-"n_singlet_cells"

qc_table(Sample_HTO_Summary)

```  
  
[Return to Contents](#tcr_hashing_top)




<a id= "tcr_well_hto_barcode"></a>  


#### scTCR Well Counts per HTO Barcode Plot

```{r sctcr_Count_By_HTO_Barcode_Plot, results="asis"}
subchunkify <- function(g, fig_height=7, fig_width=5) {
  g_deparsed <- paste0(deparse(
    function() {g}
  ), collapse = '')
  
  sub_chunk <- paste0("
  `","``{r sub_chunk_", floor(runif(1) * 10000), ", fig.height=", fig_height, ", fig.width=", fig_width, ", echo=FALSE}",
  "\n(", 
    g_deparsed
    , ")()",
  "\n`","``
  ")
  
  cat(knitr::knit(text = knitr::knit_expand(text = sub_chunk), quiet = TRUE))
}

plot_list <- list()
for (i in seq_along(pools)){
  plot_list[[i]] <- qc_aligned_barplot_facet(meta = unqiue_dt%>% filter (pool_id==pools[i]),
                   category_x = "sample_hto_pool",
                   category_y = "well_id",
                   category_name = "Well ID",
                   name_x = "HTO/Sample",
                   colorset_y = "varibow",
                   name_y = "Number of Cells",
                   facet_formula = formula("~pool_id"), nrow = 1, scales ="free_x", drop = TRUE) 
}

subchunkify(plot_list[[1]],  ceiling(n_pools/2)*4 + 0.4,max(0.6*n_wells + 1*n_pools + 3, 8))
if (length(plot_list)==2){


subchunkify(plot_list[[2]],  ceiling(n_pools/2)*4 + 0.4,max(0.6*n_wells + 1*n_pools + 3, 8))


}
cat("\n[Return to Contents](#tcr_hashing_top)", labels = "", sep = "\n")


``` 

<a id= "tcr_well_fraction_hto_barcode"></a>  

#### scTCR Well Fraction per HTO Barcode Plot


```{r tcr_hto_bar_plot, results="asis"}


options(knitr.duplicate.label = "allow")

bar_g <- qc_stacked_barplot_facet(meta = unqiue_dt,
                   category_x = "sample_hto_pool",
                   category_y = "well_id",
                   category_name = "Well ID",
                   name_x = "HTO/Sample",
                   colorset_y = "varibow",
                   name_y = "Fraction of Cells",
                   as_fraction = TRUE , 
                   facet_formula = formula("~pool_id"), nrow = 1, scales ="free_x", drop = TRUE)
                   
subchunkify(bar_g,  ceiling(n_pools/2)*4 + 0.4,max(0.6*n_wells + 1*n_pools + 3, 8))

cat("\n[Return to Contents](#tcr_hashing_top)", labels = "", sep = "\n")


``` 

<a id= "sample_scTCR_summary"></a>  


### scTCR Sample Clonotype Summary

<a id= "tcr_sample_clone_size"></a>  


#### scTCR Clonotype Size

```{r tcr_sample_clone_size_plots, results = "asis", fig.height = 3.5}
stm("Generating sample clone size plots")


dt_clone_size<-dt %>%  group_by ( sample_hto_pool, well_id,barcode) %>%   
        tally()%>%  group_by (sample_hto_pool, n)%>% mutate_if(is.numeric,as.character)
colnames(dt_clone_size)[4]<-"Clone_Size"
dt_clone_size$Clone_Size<-as.character(dt_clone_size$Clone_Size)
pool_id<-substring(gsub("^.*\\P","P",dt_clone_size$sample_hto_pool), 1,2)
dt_clone_size$pool_id<-pool_id 

ggplot(data = dt_clone_size,aes(x =sample_hto_pool, fill = factor(Clone_Size, levels=unique(dt_clone_size$Clone_Size)[order(nchar(unique(dt_clone_size$Clone_Size)), unique(dt_clone_size$Clone_Size),decreasing = TRUE)])))  +
  geom_bar( width = 0.8,position = "fill")+
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))+ylab("Percentage")+ scale_fill_discrete(name = "Clone Size")+
    facet_wrap(~pool_id, ncol = n_distinct(dt_clone_size$pool_id), scales = "free_x", drop = TRUE)
            
cat("\n[Return to Contents](#tcr_hashing_top)", labels = "", sep = "\n")


``` 





<a id= "unique_clone_type"></a>

#### scTCR Sample Unique Clonotype

```{r tcr_sample_clone_unique_plots, results = "asis", fig.height = 3.5}
stm("Generating unique clone type plots")

dt_clone_unique<-dt %>%  
        group_by ( Well_ID,sample_hto_pool, raw_clonotype_id) %>%   
        tally()            %>%  
        group_by ( sample_hto_pool,Well_ID)%>%    
        tally()
pool_id<-substring(gsub("^.*\\P","P",dt_clone_unique$Well_ID), 1,2)
dt_clone_unique$pool_id<-pool_id 
ggplot(data=dt_clone_unique, aes(x=sample_hto_pool, y=n)) +
  geom_bar(stat="identity", fill="steelblue")+   
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))+
  ylab("Number of Unique Clone Type ")+
  facet_wrap(~pool_id, ncol = n_distinct(dt_clone_unique$pool_id), scales = "free_x", drop = TRUE)

    
cat("\n[Return to Contents](#tcr_hashing_top)", labels = "", sep = "\n")


```  

<a id= "sample_cdr3_nt_distribution"></a>  

#### scTCR Sample CDR3 Nucleotide Distribution Summary

```{r tcr_sample_cdr_nt_size_plots, results = "asis", fig.height = 3.5}
stm("Generating cdr3 nt length distribution plots")

dt$cdr3_nt_length <- str_length(dt$cdr3_nt) 
pool_id<-substring(gsub("^.*\\P","P",dt$Well_ID), 1,2)
dt$pool_id<-pool_id 

ggplot(data = dt, aes(x = cdr3_nt_length, fill = sample_hto_pool)) +
    geom_bar(bin=70,position = "dodge")+ 
    xlim(10, 80)+
    facet_wrap(~pool_id, drop = TRUE)

    
cat("\n[Return to Contents](#tcr_well_top)", labels = "", sep = "\n")


``` 




<a id= "tcr_sample_clonotype_abundance"></a>  


#### scTCR Sample Clonotyoe Abundance


```{r tcr_sample_abundance_plots, results = "asis", fig.height = 3.5,fig.width =12 }
stm("Generating abundance plots")

abundance<-dt   %>%    group_by ( Well_ID,sample_hto_pool, barcode,raw_clonotype_id) %>%   
        tally() %>%    group_by ( Well_ID,sample_hto_pool,raw_clonotype_id) %>%   
        tally() %>%    group_by ( n,sample_hto_pool)%>%    
        tally()

pool_id<-substring(gsub("^.*\\P","P",abundance$sample_hto_pool), 1,2)
abundance$pool_id<-pool_id 

ggplot(abundance, aes(x = n,y = nn,color = sample_hto_pool)) +  
    geom_line()+ 
    scale_x_log10(breaks = c(1,10,100,1000))+ 
    scale_y_log10(breaks = c(1,10,100,1000,10000))+
    ylab("Number of Clonotyoe")+
    xlab("Abundance")+
    facet_wrap(~pool_id, nrow = n_distinct(abundance$pool_id), scales = "free_x", drop = TRUE)

cat("\n[Return to Contents](#tcr_hashing_top)", labels = "", sep = "\n")


```  

[Return to Contents](#tcr_hashing_top)

---  

scTCR seq report sample module v.`r {tcr_sample_module_version}`, Qiuyu Gong; Lauren Okada  


```{r sctcr_seq_sample_cleanup, include = FALSE}
module_vars <- setdiff(ls(), orig_workspace)
rm(list=module_vars)

gc()

```